### PR TITLE
Add TPM clock increment example

### PR DESCRIPTION
**Pull request by tomoveu**
New example demonstrating how to update(increment) the TPM clock.

This is the relative (uptime) clock of the TPM.

There are three specifics:
1. It can only be incremented forward
2. OWNER or Platform Authorization is needed. My example uses OWNER authorization as typically Platform auth is likely to be locked or not set by an end user.
3. If the new value is bigger than the update interval of the TPM, the TPM will internally update its non-volatile register as well.

Currently, the example increments the clock by 50 seconds (50000ms), but I think it is better to have an argument for the user.

./example/clock/clockSet {new clock time in ms}
./example/clock/clockSet without argument increments with 50 seconds as a test and example

I will add the argument option tomorrow and README with the specifics of the example :) 

Signed-off-by: Dimitar Tomov <dimi@designfirst.ee>

### CODE REVIEW

**Comment by tomoveu** [comment]
Sample output
```

~/wolfTPM$ sudo examples/clock/clockSet  
TPM2 Demo of setting the TPM clock forward
wolfTPM2_Init: success
TPM2_ReadClock: success
TPM2_ReadClock: (total)  time=23213283
TPM2_ReadClock: (uptime) clock=157963516
TPM2_clockSet: success
TPM2_ReadClock: success
TPM2_ReadClock: (total)  time=23213291
TPM2_ReadClock: (uptime) clock=158013521

	 oldClock=157963516 
	 newClock=158013521 


```

Perhaps there are too many print messages for such a demo, but I follow the established prints from the other examples.

**Comment by tomoveu** [comment]
Added optional argument for the user. Updated the examples README. Also, made some prints available only for DEBUG, so the example output is shorter and clearer. Usage prints I think turn out quite nice considering the specifics.

Sample output below

- with DEBUG enabled
```

$ sudo examples/clock/clockSet
TPM2 Demo of setting the TPM clock forward
wolfTPM2_Init: success
TPM2_ReadClock: success
TPM2_ReadClock: (uptime) time=1110056
TPM2_ReadClock: (total)  clock=170074322
TPM2_ClockSet: success
TPM2_ReadClock: success
TPM2_ReadClock: (uptime) time=1110064
TPM2_ReadClock: (total)  clock=170124327

	 oldClock=170074322 
	 newClock=170124327 
```

- without DEBUG enabled

```
~/wolfTPM$ sudo examples/clock/clockSet 200000000
TPM2 Demo of setting the TPM clock forward
wolfTPM2_Init: success
TPM2_ReadClock: success
TPM2_ClockSet: success
TPM2_ReadClock: success

	 oldClock=170195294 
	 newClock=200000005 

~/wolfTPM$ sudo examples/clock/clockSet
TPM2 Demo of setting the TPM clock forward
wolfTPM2_Init: success
TPM2_ReadClock: success
TPM2_ClockSet: success
TPM2_ReadClock: success

	 oldClock=200009115 
	 newClock=200059120 

```

- with wrong arguments

```

~/wolfTPM$ sudo examples/clock/clockSet arg1 arg2
Incorrect arguments
Expected usage:
./examples/clock/clockSet [time]
* time is value in miliseconds
	The TPM clock can be set only forward.
	The TPM clock shows the total time the TPM was ever powered.
	The TPM clock is different and always higher than the
	current uptime. The TPM uptime can not be tampered with.
Demo usage without parameters:
* increments the TPM clock with 50 seconds (50,000 ms)

```

@dgarske turning to you for review :)

File: examples/timestamp/clock_set.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+{
+    printf("Expected usage:\n");
+    printf("./examples/clock/clockSet [time]\n");
+    printf("* time is value in miliseconds\n");
+    printf("\tThe TPM clock can be set only forward.\n");
+    printf("\tThe TPM clock shows the total time the TPM was ever powered.\n");
+    printf("\tThe TPM clock is different and always higher than the\n");
+    printf("\tcurrent uptime. The TPM uptime can not be tampered with.\n");
+    printf("Demo usage without parameters:\n");
+    printf("* increments the TPM clock with 50 seconds (50,000 ms)\n");
```
Comment (embhorn) [inline examples/timestamp/clock_set.c:49]: Could just say default time is 50,000 at line 43

File: examples/timestamp/clock_set.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    UINT64 oldClock, newClock;
+
+    if (argc == 2) {
+        newClock = atoi(argv[1]);
+    }
+    else if (argc == 1) {
+        newClock = 0;
+    }
+    else {
+        printf("Incorrect arguments\n");
```
Comment (embhorn) [inline examples/timestamp/clock_set.c:77]: The only way to see the usage is to enter more than 1 args. Could you add a way to parse `?` to show the usage?

File: examples/README.md [hunk 3]
```diff
@@ -137,6 +137,22 @@ With browsers you will get certificate warnings until you load the test CA's `./
...
+
+The TPM has internal hardware clock that can be useful to the user. There are two values that the TPM can provide in respect to time.
+
+TPM time is the current uptime, since the last power on sequence. This value can not be changed or modified. There is no mechanism for that. The value is reset at every power sequence.
+
+TPM clock is the total time the TPM has ever been powered. This value can be modified using the TPM2_ClockSet command. The TPM clock can be set only forward.
+
+This way the user can keep track of relative and current time using the TPM clock.
+
+Note: If the new time value makes a change bigger than the TPM clock update interval, then the TPM will first update its volatile register for time and then the non-volatile register for time. This may cause a narrow delay before the commands returns execution to the user. Depending on the TPM manufacturer, the delay can vary from us to few ms.
```
Comment (embhorn) [inline examples/README.md:152]: Mention explicitly that there is an optional time argument to the example.

**Review by embhorn — CHANGES_REQUESTED** [review]
Very cool example! I requested a couple of tings to clarify the use of the optional time argument.

File: examples/timestamp/clock_set.c (hunk 2 above)
Comment (tomoveu) [reply examples/timestamp/clock_set.c:77]: Ah, yes. I forgot to add that option. We have something similar in the other examples. I think it was -h @dgarske ? Will check and add.

File: examples/README.md (hunk 3 above)
Comment (tomoveu) [reply examples/README.md:152]: Will do.

**Comment by tomoveu** [comment]
@embhorn feedback taken in, please re-review ;)

**Review by embhorn — APPROVED** [review]
Looks great! Thanks, @tomoveu !

**Review by dgarske — APPROVED** [review]

**Pull request merged**